lib/httplib/connstate.go

```diff
@@ -0,0 +1,221 @@
+// Its Notify method implements the http.Server.ConnState function
+// which should be attached to the Server instance to initiate tracking.
+// Calling Channel and passing a connection will return a chan that
+// receives state change events.
+//
+// All net.Conn instances passed to this type must be comparable or calls
+// will panic. The net.Conn instances passed by http.Server such as *tls.Conn
+// are comparable.
+//
+//    server := http.Server{...}
+//    connState := NewServerConnState()
+//    server.ConnState = connState.Notify
+//    ...
+//    tlsConn := tls.Server(...)
+//    listener := listener.Wrap(tlsConn) // fake listener to pass to Serve
+//    ch := connState.Channel(tlsConn, http.StateClosed)
+//    server.Serve(listener)
+//    <-ch // releases once tlsConn is closed
+type ServerConnState struct {
+	mu sync.Mutex
+
+	// conns tracks open connection state. A ConnStateEvents is created
+	// for each connection, which processes notifications for subscribers
+	// of that connection.
+	//
+	// NOTE: using net.Conn as a map key is not safe if someone creates
+	// a net.Conn using a concrete type that is not comparable. net.Conn
+	// is stateful, so implementations should be comparable. Using net.Conn
+	// simplified the implementation and allows for this functionality to
+	// be extended, at the expense of strange implementations that will
+	// panic on the first test.
+	conns map[net.Conn]*ConnStateEvents
+}
+
+// NewServerConnState returns an initialized ServerConnState.
+func NewServerConnState() *ServerConnState {
+	return &ServerConnState{conns: make(map[net.Conn]*ConnStateEvents)}
+}
+
+// Notify subscribers of a connection state change.
+// Terminal states remove the connection from tracking.
+//
+// Notify implements http.Server.ConnState.
+func (s *ServerConnState) Notify(conn net.Conn, state http.ConnState) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	c, ok := s.conns[conn]
+	if TerminalConnState(state) {
+		delete(s.conns, conn)
+	} else if !ok {
+		s.conns[conn] = nil
+	}
+
+	if c != nil {
+		c.Notify(state)
+	}
+}
+
+// Channel returns a read-only channel that will receive http.ConnState
+// events for conn. The state parameter determines which states are
+// received by the channel. The channel will receive all events when state
+// is AllConnStates. All channels will receive terminal states (http.StateClosed
+// & http.StateHijacked) events after which the channel is closed. Multiple
+// calls for the same connection are supported.
+//
+// Passing a closed connection will result in no events received and leak
+// both the conn and channel unless followed by a Release call.
+//
+// Calling Channel when conn is already in the desired state will result
+// in no events until the connection moves to a different state and back
+// again (e.g. idle->active->idle).
+func (s *ServerConnState) Channel(conn net.Conn, state http.ConnState) <-chan http.ConnState {
```

Comment:
Rename this function to something a bit more indicative of what it does and not what it returns? Something like `WatchForState` or `SubscribeForState`?

Reply:
I renamed this a few times, but stuck with channel because it returns a channel. It was previously Subscribe, but I thought that would be too generic and conflict if we wanted to add another method that notified by callback instead. `ForState` is more descriptive, but this utility's only function is to emit state.